scipy/spatial/distance.py

```diff
@@ -1069,77 +1154,37 @@ def sokalsneath(u, v):
 _SIMPLE_CDIST = {}
 _SIMPLE_PDIST = {}
 
-for names, wrap_name in [
-    (['braycurtis'], "bray_curtis"),
-    (['canberra'], "canberra"),
-    (['chebychev', 'chebyshev', 'cheby', 'cheb', 'ch'], "chebyshev"),
-    (["cityblock", "cblock", "cb", "c"], "city_block"),
-    (["euclidean", "euclid", "eu", "e"], "euclidean"),
-    (["sqeuclidean", "sqe", "sqeuclid"], "sqeuclidean"),
+for wrap_name, names, typ in [
+    ("bray_curtis", ['braycurtis'], "double"),
+    ("canberra", ['canberra'], "double"),
+    ("chebyshev", ['chebychev', 'chebyshev', 'cheby', 'cheb', 'ch'], "double"),
+    ("city_block", ["cityblock", "cblock", "cb", "c"], "double"),
+    ("euclidean", ["euclidean", "euclid", "eu", "e"], "double"),
+    ("sqeuclidean", ["sqeuclidean", "sqe", "sqeuclid"], "double"),
+    ("dice", ["dice"], "bool"),
+    ("kulsinski", ["kulsinski"], "bool"),
+    ("hamming", ["matching"], "bool"),
+    ("rogerstanimoto", ["rogerstanimoto"], "bool"),
+    ("russellrao", ["russellrao"], "bool"),
+    ("sokalmichener", ["sokalmichener"], "bool"),
+    ("sokalsneath", ["sokalsneath"], "bool"),
+    ("yule", ["yule"], "bool"),
 ]:
-    cdist_fn = getattr(_distance_wrap, "cdist_%s_wrap" % wrap_name)
-    pdist_fn = getattr(_distance_wrap, "pdist_%s_wrap" % wrap_name)
+    converter = {"bool": _convert_to_bool,
+                 "double": _convert_to_double}[typ]
+    fn_name = {"bool": "%s_bool_wrap",
+               "double": "%s_wrap"}[typ] % wrap_name
+    cdist_fn = getattr(_distance_wrap, "cdist_%s" % fn_name)
+    pdist_fn = getattr(_distance_wrap, "pdist_%s" % fn_name)
     for name in names:
-        _SIMPLE_CDIST[name] = _convert_to_double, cdist_fn
-        _SIMPLE_PDIST[name] = _convert_to_double, pdist_fn
-
-for name in ["dice", "kulsinski", "matching", "rogerstanimoto", "russellrao",
-             "sokalmichener", "sokalsneath", "yule"]:
-    wrap_name = "hamming" if name == "matching" else name
-
-    cdist_fn = getattr(_distance_wrap, "cdist_%s_bool_wrap" % wrap_name)
-    _SIMPLE_CDIST[name] = _convert_to_bool, cdist_fn
-
-    pdist_fn = getattr(_distance_wrap, "pdist_%s_bool_wrap" % wrap_name)
-    _SIMPLE_PDIST[name] = _convert_to_bool, pdist_fn
+        _SIMPLE_CDIST[name] = converter, cdist_fn
+        _SIMPLE_PDIST[name] = converter, pdist_fn
```

Comment:
Reorganising this portion of code is something I was thinking too while working at #6656. 
Anyway, IMHO, I would rather probably do this in a standalone PR since it is quite independent from weights.
Plus, it is probably worth doing something more generic that takes care also of doing the right type conversions for metrics that supports more types or have more than one wrapper.